File: aviloo-mcp/src/aviloo_mcp/repo.py
"""Resolve and constrain paths to the configured project repository."""

from __future__ import annotations

import os
import re
from pathlib import Path

_HEX32 = re.compile(r"^[0-9a-fA-F]{32}$")
_UUID = re.compile(
    r"^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$"
)

_PACKAGE_DIR = Path(__file__).resolve().parent
_DEFAULT_REPO_ROOT = _PACKAGE_DIR.parent.parent.parent.parent


def repo_root() -> Path:
    env = os.environ.get("AVILOO_MCP_REPO_ROOT")
    if env:
        return Path(env).resolve()
    return _DEFAULT_REPO_ROOT.resolve()


def assert_in_repo(path: Path) -> Path:
    """Resolve path and ensure it stays under REPO_ROOT."""
    root = repo_root()
    resolved = path.resolve()
    try:
        resolved.relative_to(root)
    except ValueError as exc:
        raise ValueError(f"Path is outside repository: {resolved}") from exc
    return resolved


def _looks_like_aviloo_pdf(path: Path) -> bool:
    name_lower = path.name.lower()
    if "aviloo" in name_lower:
        return True
    if _HEX32.match(path.stem):
        return True
    parts = path.parts
    if "docs" in parts and path.suffix.lower() == ".pdf":
        return True
    if path.parent.resolve() == repo_root() and path.suffix.lower() == ".pdf":
        return True
    return False


def find_pdfs() -> list[Path]:
    """List PDF paths in the repo that may be AVILOO certificates."""
    root = repo_root()
    if not root.is_dir():
        return []
    candidates: list[Path] = []
    for path in root.rglob("*.pdf"):
        if not path.is_file():
            continue
        try:
            assert_in_repo(path)
        except ValueError:
            continue
        if _looks_like_aviloo_pdf(path):
            candidates.append(path)
    return sorted(candidates, key=lambda p: str(p.relative_to(root)))


def find_cert_in_repo(cert_or_id: str) -> Path | None:
    """Find PDF by certificate UUID or 32-char hex download id (filename stem)."""
    key = cert_or_id.strip()
    key_lower = key.lower()
    root = repo_root()

    if _HEX32.match(key):
        for pdf in find_pdfs():
            if pdf.stem.lower() == key_lower:
                return pdf
        direct = root / f"{key_lower}.pdf"
        if direct.is_file():
            return assert_in_repo(direct)
        for pdf in root.rglob(f"{key_lower}.pdf"):
            try:
                return assert_in_repo(pdf)
            except ValueError:
                continue

    if _UUID.match(key):
        from aviloo_mcp.parser import extract_text, parse_aviloo_text

        target = key.upper()
        for pdf in find_pdfs():
            if pdf.stem.upper() == target:
                return pdf
        for pdf in find_pdfs():
            try:
                text = extract_text(pdf)
                parsed = parse_aviloo_text(text, download_id=pdf.stem)
                cert = parsed.get("certificate_number")
                if cert and cert.upper() == target:
                    return pdf
            except Exception:
                continue
    return None

File: internal/aviloo/repo/repo.go
package repo

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

var (
	hex32 = regexp.MustCompile(`^[0-9a-fA-F]{32}$`)
	uuid  = regexp.MustCompile(`^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}$`)
)

// Root returns the configured repository root.
func Root() (string, error) {
	if env := os.Getenv("AVILOO_MCP_REPO_ROOT"); env != "" {
		return filepath.Abs(env)
	}
	return os.Getwd()
}

// AssertInRepo resolves path and ensures it stays under the repo root.
func AssertInRepo(path string) (string, error) {
	root, err := Root()
	if err != nil {
		return "", err
	}
	abs, err := filepath.Abs(path)
	if err != nil {
		return "", err
	}
	rel, err := filepath.Rel(root, abs)
	if err != nil || strings.HasPrefix(rel, "..") {
		return "", fmt.Errorf("path is outside repository: %s", abs)
	}
	return abs, nil
}

func looksLikeAvilooPDF(root, abs string) bool {
	nameLower := strings.ToLower(filepath.Base(abs))
	if strings.Contains(nameLower, "aviloo") {
		return true
	}
	stem := strings.TrimSuffix(nameLower, filepath.Ext(nameLower))
	if hex32.MatchString(stem) {
		return true
	}
	rel, err := filepath.Rel(root, abs)
	if err != nil {
		return false
	}
	parts := strings.Split(rel, string(filepath.Separator))
	for _, part := range parts {
		if part == "docs" && strings.HasSuffix(strings.ToLower(abs), ".pdf") {
			return true
		}
	}
	if filepath.Dir(abs) == root && strings.HasSuffix(strings.ToLower(abs), ".pdf") {
		return true
	}
	return false
}

// FindPDFs lists PDF paths in the repo that may be AVILOO certificates.
func FindPDFs() ([]string, error) {
	root, err := Root()
	if err != nil {
		return nil, err
	}
	info, err := os.Stat(root)
	if err != nil || !info.IsDir() {
		return nil, nil
	}
	var candidates []string
	err = filepath.WalkDir(root, func(path string, d os.DirEntry, walkErr error) error {
		if walkErr != nil {
			return nil
		}
		if d.IsDir() || !strings.HasSuffix(strings.ToLower(path), ".pdf") {
			return nil
		}
		if _, err := AssertInRepo(path); err != nil {
			return nil
		}
		if looksLikeAvilooPDF(root, path) {
			candidates = append(candidates, path)
		}
		return nil
	})
	if err != nil {
		return nil, err
	}
	sort.Slice(candidates, func(i, j int) bool {
		ri, _ := filepath.Rel(root, candidates[i])
		rj, _ := filepath.Rel(root, candidates[j])
		return ri < rj
	})
	return candidates, nil
}

// FindPDFsForRegnr returns PDFs whose path contains the registration number.
func FindPDFsForRegnr(regnr string) ([]string, error) {
	regnr = strings.ToUpper(strings.TrimSpace(regnr))
	if regnr == "" {
		return nil, nil
	}
	pdfs, err := FindPDFs()
	if err != nil {
		return nil, err
	}
	var matches []string
	for _, pdf := range pdfs {
		rel := strings.ToUpper(Rel(pdf))
		if strings.Contains(rel, regnr) || strings.Contains(strings.ToUpper(filepath.Base(pdf)), regnr) {
			matches = append(matches, pdf)
		}
	}
	return matches, nil
}

// Rel returns a path relative to the repo root when possible.
func Rel(abs string) string {
	root, err := Root()
	if err != nil {
		return abs
	}
	rel, err := filepath.Rel(root, abs)
	if err != nil {
		return abs
	}
	return rel
}

// CertFinder resolves certificate PDFs by UUID or hex download id.
type CertFinder interface {
	FindCertInRepo(certOrID string) (string, error)
}

// FindCertByStem looks up a PDF by 32-char hex filename stem.
func FindCertByStem(root, key string) (string, error) {
	pdfs, err := FindPDFs()
	if err != nil {
		return "", err
	}
	keyLower := strings.ToLower(key)
	for _, pdf := range pdfs {
		stem := strings.TrimSuffix(filepath.Base(pdf), filepath.Ext(pdf))
		if strings.ToLower(stem) == keyLower {
			return pdf, nil
		}
	}
	direct := filepath.Join(root, keyLower+".pdf")
	if info, err := os.Stat(direct); err == nil && !info.IsDir() {
		return AssertInRepo(direct)
	}
	var matches []string
	_ = filepath.WalkDir(root, func(path string, d os.DirEntry, walkErr error) error {
		if walkErr != nil || d.IsDir() {
			return nil
		}
		if strings.EqualFold(filepath.Base(path), keyLower+".pdf") {
			if abs, err := AssertInRepo(path); err == nil {
				matches = append(matches, abs)
			}
		}
		return nil
	})
	if len(matches) > 0 {
		return matches[0], nil
	}
	return "", nil
}
